src/dpl/src/Place.cpp

```diff
@@ -796,58 +793,97 @@ PixelPt Opendp::diamondSearch(const Cell* cell,
-          && abs(y_offset) < scaled_max_displacement_y_) {
-        if (j % 2 == 1) {
-          y_offset = -y_offset;
-        }
-        diamondSearchSide(cell,
-                          x,
-                          y,
-                          x_min,
-                          y_min,
-                          x_max,
-                          y_max,
-                          x_offset,
-                          y_offset,
-                          best_pt,
-                          best_dist);
-      }
+  struct PQ_entry
+  {
+    int manhattan_distance;
+    odb::Point p;
+    bool operator>(const PQ_entry& other) const
+    {
+      return manhattan_distance > other.manhattan_distance;
+    }
+    bool operator==(const PQ_entry& other) const
+    {
+      return manhattan_distance == other.manhattan_distance;
+    }
+  };
+  std::priority_queue<PQ_entry, std::vector<PQ_entry>, std::greater<PQ_entry>>
+      positionsHeap;
+
+  std::unordered_set<odb::Point, odb::Point::Hash> visitedPositions;
+  const odb::Point center = {x, y};
+  positionsHeap.push({0, center});
+  PixelPt best_pt;
+  int best_dist = std::numeric_limits<int>::max();
+  int best_for = 0;
+  while (!positionsHeap.empty()) {
+    auto top = positionsHeap.top();
+    positionsHeap.pop();
+    const int new_x = top.p.getX();
+    const int new_y = top.p.getY();
+    const int x_offset = new_x - x;
+    const int y_offset = new_y - y;
+    // Check if this position has been visited
+    if (visitedPositions.count({new_x, new_y}) > 0) {
+      continue;
     }
 
-    // right side
-    for (int j = 1; j < (i + 1) * 2; j++) {
-      int x_offset = (j - 1) / 2;
-      int y_offset = ((i + 1) * 2 - j) / 2;
-      if (abs(x_offset) < max_displacement_x_
-          && abs(y_offset) < scaled_max_displacement_y_) {
-        if (j % 2 == 1) {
-          y_offset = -y_offset;
-        }
-        diamondSearchSide(cell,
-                          x,
-                          y,
-                          x_min,
-                          y_min,
-                          x_max,
-                          y_max,
-                          x_offset,
-                          y_offset,
-                          best_pt,
-                          best_dist);
+    PixelPt cur_pt;
+    int cur_dist = 0;
+    if (abs(x_offset) < max_displacement_x_
+        && abs(y_offset) < scaled_max_displacement_y_) {
+      diamondSearchSide(cell,
+                        x,
+                        y,
+                        x_min,
+                        y_min,
+                        x_max,
+                        y_max,
+                        x_offset,
+                        y_offset,
+                        cur_pt,
+                        cur_dist);
+
+      if (cur_pt.pixel && cur_dist < best_dist) {
+        best_pt = cur_pt;
+        best_dist = cur_dist;
+        best_for = 0;
+      } else if (best_dist < cur_dist) {
+        ++best_for;
       }
+      if (best_for >= 7) {
```

Comment:
What is 7?

Reply:
7 is a magic number that I found from experimenting. When you find an available point, you could return immediately or you can wait until you are sure that it was indeed the best after you do some more search.

7 means that your point remained the best you could find for 7 additional searches. any number would work, but numbers below 7 result in worse metrics, and above 7 doesn't improve.

of course it depends on the test case, but remember that everytime you find a better point, you still need to make sure it will remain the "best for" X additional searches, so you reset the counter.

Larger numbers slow down runtime with no additional gains

Reply:
If it's a magic number then it should probably be changed to a self-documenting variable and included with the others as `static constexpr`

Reply:
How would you get a better result later - aren't we expanding outwards?